sql/wsrep_high_priority_service.cc

```diff
@@ -462,6 +463,14 @@ int Wsrep_high_priority_service::log_dummy_write_set(const wsrep::ws_handle& ws_
       cs.before_rollback();
       cs.after_rollback();
     }
+
+    if (!WSREP_EMULATE_BINLOG(m_thd))
+    {
+        Wsrep_server_state::instance().provider().wait_for_gtid(
```

Comment:
Could this be replaced by hooking with MariaDB native group commit? The following patch would register the XID sync for group commit if binlog is enabled. This would allow releasing provider side commit order immediately.
```
diff --git a/sql/wsrep_binlog.cc b/sql/wsrep_binlog.cc
index da899321ba8..8cd1172b31b 100644
--- a/sql/wsrep_binlog.cc
+++ b/sql/wsrep_binlog.cc
@@ -358,7 +358,7 @@ void wsrep_register_for_group_commit(THD *thd)
     DBUG_VOID_RETURN;
   }
 
-  DBUG_ASSERT(thd->wsrep_trx().state() == wsrep::transaction::s_committing);
+  DBUG_ASSERT(thd->wsrep_trx().ordered());
 
   wait_for_commit *wfc= thd->wait_for_commit_ptr= &thd->wsrep_wfc;
 
@@ -381,9 +381,8 @@ void wsrep_register_for_group_commit(THD *thd)
 
 void wsrep_unregister_from_group_commit(THD *thd)
 {
-  DBUG_ASSERT(thd->wsrep_trx().state() == wsrep::transaction::s_ordered_commit||
-              // ordered_commit() failure results in s_aborting state
-              thd->wsrep_trx().state() == wsrep::transaction::s_aborting);
+  DBUG_ASSERT(thd->wsrep_trx().ordered());
+
   wait_for_commit *wfc= thd->wait_for_commit_ptr;
 
   if (wfc)
diff --git a/sql/wsrep_high_priority_service.cc b/sql/wsrep_high_priority_service.cc
index 7d333be1654..4d6395a557e 100644
--- a/sql/wsrep_high_priority_service.cc
+++ b/sql/wsrep_high_priority_service.cc
@@ -466,13 +466,21 @@ int Wsrep_high_priority_service::log_dummy_write_set(const wsrep::ws_handle& ws_
 
     if (!WSREP_EMULATE_BINLOG(m_thd))
     {
-        Wsrep_server_state::instance().provider().wait_for_gtid(
-            wsrep::gtid(ws_meta.group_id(),
-            wsrep::seqno(ws_meta.seqno().get() - 1)), 10);
+      wsrep_register_for_group_commit(m_thd);
+      ret = ret || cs.provider().commit_order_leave(ws_handle, ws_meta, err);
+      m_thd->wait_for_prior_commit();
     }
 
     wsrep_set_SE_checkpoint(ws_meta.gtid(), wsrep_gtid_server.gtid());
-    ret= ret || cs.provider().commit_order_leave(ws_handle, ws_meta, err);
+
+    if (!WSREP_EMULATE_BINLOG(m_thd))
+    {
+      wsrep_unregister_from_group_commit(m_thd);
+    }
+    else
+    {
+      ret= ret || cs.provider().commit_order_leave(ws_handle, ws_meta, err);
+    }
     cs.after_applying();
   }
   DBUG_RETURN(ret);
```